messaging/src/main/java/org/axonframework/messaging/QueueMessageStream.java

```diff
@@ -0,0 +1,171 @@
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    /**
+     * Constructs an instance with an unbounded queue. Offering elements will always be possible, as long as memory
+     * permits.
+     */
+    public QueueMessageStream() {
+        this(new LinkedBlockingQueue<>());
+    }
+
+    /**
+     * Construct an instance with given {@code queue} as the underlying queue. Offering and consuming items will depend
+     * on the semantics of the implementation of the queue.
+     * <p>
+     * Note that delivery and consumption of items is done through {@link BlockingQueue#offer(Object)} and
+     * {@link BlockingQueue#poll()}, respectively. This means that a queue must be available to buffer elements.
+     * Implementations of a {@link java.util.concurrent.TransferQueue} typically don't have this, and will therefore not
+     * work.
+     *
+     * @param queue The queue to use to store messages in transit from producer to consumer
+     */
+    public QueueMessageStream(BlockingQueue<Entry<M>> queue) {
+        this.queue = queue;
+    }
+
+    /**
+     * Add the given {@code message} and accompanying {@code context} available for reading by a consumer. Any callback
+     * that has been registered with be notified of the availability of a new element.
+     * <p>
+     * If the underling buffer has insufficient space to store the offered element, the method returns {@code false}.
+     *
+     * @param message The message to add to the queue
+     * @param context The context to accompany the message
+     * @return {@code true} if the message was successfully buffered. Otherwise {@code false}
+     */
+    public boolean offer(@Nonnull M message, @Nonnull Context context) {
+        if (queue.offer(new SimpleEntry<>(message, context))) {
+            onAvailableCallbackRef.get().run();
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Marks the queue as completed, indicating to any consumer that no more messages will become available.
+     * <p>
+     * Note that there is no validation on offering items whether the stream is completed. It is the caller's
+     * responsibility to ensure no Messages are {@link #offer(Message, Context) offered} after completion.
+     */
+    public void complete() {
+        completed.set(true);
+        onAvailableCallbackRef.get().run();
+    }
+
+    /**
+     * Marks the queue as completed exceptionally caused by given {@code error}, indicating to any consumer that no more
+     * messages will become available.
+     * <p>
+     * Note that there is no validation on offering items whether the stream is completed. It is the caller's
+     * responsibility to ensure no Messages are {@link #offer(Message, Context) offered} after completion.
+     *
+     * @param error The cause of the exceptional completion
+     */
+    public void complete(Throwable error) {
```

Comment:
Curious: Wouldn't calling this `completeExceptionally` align more with other  APIs out there? Or do you feel that's redundant given the parameter name?